# Momma Tables.where() and all her little predicates

Filtering datatables is done withe the .where() method for which there are a whole slew of "predicates" that indicate the filtering condition. As always, you learn best by practicing.

In [98]:
# Load the required modules
import numpy as np
from datascience import * 

## Philadelphia Services
To keep things interesting, we'll take a look at a dataset available on [OpenDataPhilly](https://opendataphilly.org/datasets/) that contains services available in the city.

In [99]:
path = 'data/'
data = path + "philly_services.csv"
services = Table.read_table(data)
services

Id   | Category              | Organization Name                                 | Address                                       | Zip Code | Days                                                         | Time: Open | Time: Close | People Served              | Description                                                  | Phone Number    | LatLon
1    | Day / Drop in Centers | A New Life Recovery & Education Center            | 3119 Spring Garden St, Philadelphia, PA 19104 | 19104    | Tues-Sat: 10-6pm Recovery & peer support; 1-3pm Snacks W ... | 10PM       | 6PM         | Women, Men, Families       | Services provided: Recovery & peer support, Snacks, Laun ... | 267-507-3529    | -75.18833430000001,39.9635841
2    | Emergency Shelter     | ACTS Christian Transistional Services             | 1428 N. 28th St, Philadelphia, PA 19121       | 19121    | 2nd, 3rd, 4th Wednesdays, 6-9 PM                             | 6PM        | 9PM         | Women, Women with Children | ACTS is a recovery house for women and their children pr ... | (215) 769-9799  | -75.1812997,39.97794469999999
3    | Emergency Shelter     | Apple Tree Family Center                          | 1430 Cherry St, Philadelphia, PA 19102        | 19102    | Mon-Fri 7AM - 3PM                                            | nan        | nan         | Women, Families            | Families/Single Women Seeking Emergency Housing Services ... | 215-686-7150/52 | -75.1641606,39.9554759
4    | Emergency Food        | Grace Café at Arch Street United Methodist Church | 55 N Broad St, Philadelphia, PA 19107         | 19107    | Sunday 5:30PM - 7PM                                          | 5:30 PM    | 7:00 PM     | Women, Men, Families       | Dinner Sun: 5:30-7:00pm                                      | 215-568-6250    | -75.16290719999999,39.954096
5    | Permanent Housing     | Bethesda Bainbridge                               | 700 S. 15th St, Philadelphia, PA 19146        | 19146    | nan                                                          | nan        | nan         | Men                        | Permanent single room occupancy (SRO) housing for 30 men ... | (215) 985-1600  | -75.167497,39.9429167
6    | Permanent Housing     | Bethesda North Broad                              | 722 N. Broad St, Philadelphia, PA 19130       | 19130    | nan                                                          | nan        | nan         | Women                      | Permanent SRO housing for 49 men and women with historie ... | (215) 985-1600  | -75.1605817,39.9678332
7    | Permanent Housing     | Bethesda Spruce                                   | 1110 Spruce St Philadelphia, PA 19107         | 19107    | nan                                                          | nan        | nan         | Women                      | Permanent SRO housing for 16 women who cope with chronic ... | (215) 985-1600  | -75.16005419999999,39.94611099999999
8    | Emergency Food        | Bright Hope Baptist Church                        | 1601 N 12th St, Philadelphia, PA 19122        | 19122    | Mon - Thurs (6PM-6:45PM) & Sat (11AM-12PM)                   | 6:00 PM    | 6:45 PM     | Women, Men, Families       | Monday-Thursday meals closed July and August; clothing,  ... | 215-232-6004    | -75.1540249,39.978021
9    | Emergency Food        | Breaking Bread at Broad Street Ministry           | 315 S. Broad Street Philadelphia, PA 19107    | 19107    | Breakfast Mon: 7:45-8:45am Lunch Thurs & Fri: 11:30-12:1 ... | nan        | nan         | Women, Men, Families       | BSM’s Breaking Bread meals are open to anyone who wishes ... | 215-735-4847    | -75.1644824,39.9461067
10   | Emergency Shelter     | Philadelphia Brotherhood Rescue Mission (PBRM)    | 401 E. Girard Avenue, Philadelphia, PA 19125  | 19125    | Mon to Sun, 6:30 AM to 12:00 AM Open 24 hours for emerge ... | 24 hours   | nan         | Men                        | Established in 1908, PBRM serves the homeless, the needy ... | 866-HOPE-4

## Predicates
The .where() method of datatables takes a variety of "predicates," which are express the condition you wish to use to filter the data.

## Services in the Temple zip code

Temple's zip code ia 19122. Let's find the services within the Temple zip code.

The general format of the .where() methods is .where(column_name, predicate), so we want to filter on the 'Zip Code' column and use the predicate are.equal_to(). A new table is returned with just the rows that match our condition.

In [100]:
temple_zipcode = services.where('Zip Code', are.equal_to(19122))
temple_zipcode

Id   | Category          | Organization Name          | Address                                      | Zip Code | Days                                       | Time: Open | Time: Close | People Served        | Description                                                  | Phone Number | LatLon
8    | Emergency Food    | Bright Hope Baptist Church | 1601 N 12th St, Philadelphia, PA 19122       | 19122    | Mon - Thurs (6PM-6:45PM) & Sat (11AM-12PM) | 6:00 PM    | 6:45 PM     | Women, Men, Families | Monday-Thursday meals closed July and August; clothing,  ... | 215-232-6004 | -75.1540249,39.978021
20   | Emergency Shelter | Drueding Center            | 413 W. Master Street, Philadelphia, PA 19122 | 19122    | nan                                        | nan        | nan         | Women, Families      | As the first Philadelphia program to provide both transi ... | nan          | -75.14305809999999,39.97259469999999
50   | Emergency Food    | Mt. Olive Holy Temple      | 1469 N Broad St, Philadelphia, PA 19122      | 19122    | Tuesday                                    | 12:00 PM   | 2:00 PM     | Women, Men, Families | Lunch Tues: 12-2pm                                           | 215-765-3372 | -75.1580833,39.9756058

**Challenge:** Try finding all services with a zipcode of 19107

## Looking for Food Services
Now suppose we wanted all the services where "Food" is part of the Category column. We need a different predicate. are.containing() works when the specified column contains strings and looks to see if it contains the desired substring. To see if the text in the "Category" column contains the subtring "Food" we do this:

In [101]:
food = services.where('Category', are.containing("Food"))
food

Id   | Category       | Organization Name                                 | Address                                              | Zip Code | Days                                                         | Time: Open | Time: Close | People Served              | Description                                                  | Phone Number   | LatLon
4    | Emergency Food | Grace Café at Arch Street United Methodist Church | 55 N Broad St, Philadelphia, PA 19107                | 19107    | Sunday 5:30PM - 7PM                                          | 5:30 PM    | 7:00 PM     | Women, Men, Families       | Dinner Sun: 5:30-7:00pm                                      | 215-568-6250   | -75.16290719999999,39.954096
8    | Emergency Food | Bright Hope Baptist Church                        | 1601 N 12th St, Philadelphia, PA 19122               | 19122    | Mon - Thurs (6PM-6:45PM) & Sat (11AM-12PM)                   | 6:00 PM    | 6:45 PM     | Women, Men, Families       | Monday-Thursday meals closed July and August; clothing,  ... | 215-232-6004   | -75.1540249,39.978021
9    | Emergency Food | Breaking Bread at Broad Street Ministry           | 315 S. Broad Street Philadelphia, PA 19107           | 19107    | Breakfast Mon: 7:45-8:45am Lunch Thurs & Fri: 11:30-12:1 ... | nan        | nan         | Women, Men, Families       | BSM’s Breaking Bread meals are open to anyone who wishes ... | 215-735-4847   | -75.1644824,39.9461067
11   | Emergency Food | Philadelphia Brotherhood Rescue Mission (PBRM)    | 401 E. Girard Avenue, Philadelphia, PA 19125         | 19125    | Mon to Sun, 6:30 AM to 12:00 AM Open 24 hours for emerge ... | 6:30AM     | 12:00AM     | Men                        | Established in 1908, PBRM serves the homeless, the needy ... | 866-HOPE-4-MEN | -75.13049199999999,39.9704849
12   | Emergency Food | Chosen 300 Outreach Center                        | 1116 Spring Garden St, Philadelphia, PA 19123        | 19123    | Mon: 6:00PM, Wed: 6:00PM, Fri: 6:00PM, Sat: 5:00PM, Sun: ... | 6:00PM     | nan         | Women, Men, Families       | When the economy is impacted, companies pull back from c ... | 215-765-9806   | -75.1569384,39.961445
13   | Emergency Food | Chosen 300 Outreach Center                        | 16th & Ben Franklin Parkway, Philadelphia, PA 19103  | 19103    | Sat: 5:00PM                                                  | 5:00PM     | nan         | Women, Men, Families       | When the economy is impacted, companies pull back from c ... | nan            | -75.1663793,39.9547511
14   | Emergency Food | Chosen 300 Outreach Center                        | 3959 Lancaster Ave, Philadelphia, PA 19104           | 19104    | Sun: 10:00AM, Mon: 6:00PM, Fri: 6:00PM                       | 6:00PM     | nan         | Women, Men, Families       | When the economy is impacted, companies pull back from c ... | nan            | -75.20195199999999,39.96317
32   | Emergency Food | Helping Hand Rescue Mission (Philly Restart)      | 19th & Ben Franklin Parkway, Philadelphia, PA, 19103 | 19103    | Mon                                                          | 4:00 PM    | 5:00 PM     | Women, Men, Families       | Restart has been serving meals every Monday (4pm) on the ... | 215-627-1656   | -75.1700704,39.957253
33   | Emergency Food | Helping Hand Rescue Mission                       | 610 N. 6th Street, Philadelphia, PA 19123            | 19123    | Sun                                                          | 1:30 PM    | nan         | Women, Men, Families       | We serve a first class meal on Sunday afternoons at the  ... | nan            | -75.1479812,39.9625561
46   | Emergency Food | Mercy Hospice                                     | 334 S 13th St, Philadelphia, PA 19107                | 19107    | Lunch Mon-Fri: Women with children 11am; Single women 12 ... | nan        | nan         | Women, Women with Children | Lunch Mon-Fri: Women with children 11am; Single women 12 ... | 215-545-5153   | -75.1630715,39.94563

**Challenge:** Find the places serving food that serve children

## Chaining multiple conditions
Suppose we want to find just the food services in the Temple zipcode. You can "chain" the where conditions, applying each to the result of the previous.

In [102]:
services.where('Zip Code', are.equal_to(19122)).where('Category', are.containing("Food"))

Id   | Category       | Organization Name          | Address                                 | Zip Code | Days                                       | Time: Open | Time: Close | People Served        | Description                                                  | Phone Number | LatLon
8    | Emergency Food | Bright Hope Baptist Church | 1601 N 12th St, Philadelphia, PA 19122  | 19122    | Mon - Thurs (6PM-6:45PM) & Sat (11AM-12PM) | 6:00 PM    | 6:45 PM     | Women, Men, Families | Monday-Thursday meals closed July and August; clothing,  ... | 215-232-6004 | -75.1540249,39.978021
50   | Emergency Food | Mt. Olive Holy Temple      | 1469 N Broad St, Philadelphia, PA 19122 | 19122    | Tuesday                                    | 12:00 PM   | 2:00 PM     | Women, Men, Families | Lunch Tues: 12-2pm                                           | 215-765-3372 | -75.1580833,39.9756058

## Filtering by locations
The dataset include a column "LatLon," but we really need two separate columns: one with the latitude and one with the longitude. To split this column requires a few advanced tricks we haven't covered yet, so I'm going to just show you how I did it with some comments as explanation, but you should come back to this after we learn about how to define function.

Note: This technique may prove useful in the future is you use other datasets from OpenDataPhilly

In [105]:
# Create a function that splits a string on the comma and returns either the first or second piece.
def string_split(string, col):
    return float(string.split(',')[col])

# Vectorize the function so that we can apply it to an numpy array and it will operate on every element of the array
v_string_split = np.vectorize(string_split)

In [106]:
# Split the stings, creating two new numpy array
lat_lon = services.column('LatLon')
lon = v_string_split(lat_lon, 0)
lat = v_string_split(lat_lon, 1)

In [107]:
# We chain three methods:
# First, drop the old LatLon column, then add each of the new columns
services = services.drop('LatLon').with_column('Lat', lat).with_column('Lon', lon)

## Making a Map
Following [this example](https://inferentialthinking.com/chapters/08/5/Bike_Sharing_in_the_Bay_Area.html?highlight=map) in our textbook, let's use the new Lat Lon columns to make a map showing the locations of these services.

You can zoom in, zoom out, pan, or click on a marker to see the label.

In [112]:
# Marker.map_table(stations.select('lat', 'long', 'name').relabel('name', 'labels'))
Marker.map_table(services.select('Lat', 'Lon', 'Organization Name').relabel('Organization Name', 'labels'))

## All services North of Temple

The coordinates of the Temple Bell Tower are: 39.9813° N, 75.1544° W
Let's use the where filter to find only the services north of the bell tower.

In [115]:
bell_tower_lat = 39.9813

northern_services = services.where('Lat', are.above(bell_tower_lat))
Marker.map_table(northern_services.select('Lat', 'Lon', 'Organization Name').relabel('Organization Name', 'labels'))

**Challenge:** Plot just the sevices in the Temple zipcode.

# Summary
I hope you had fun. The where() method is not that hard once you understand the basic idea. I encourage you to explore this data set and see if you can discover any interesting patterns!